File: backend/app/plan_rules.py
# CHURVOX_LOCKED_BACKEND_PLAN_RULES_20260602
# Single backend source for Churvox plan names, pricing, caps and feature flags.
# Backend billing keys stay: solo, team, pro, enterprise.
# Customer-facing names are: Start, Crew, Operator, Command.

PLAN_ORDER = {
    "solo": 1,
    "team": 2,
    "pro": 3,
    "enterprise": 4,
}

PLAN_DISPLAY_NAMES = {
    "solo": "Start",
    "team": "Crew",
    "pro": "Operator",
    "enterprise": "Command",
}

PLAN_FEATURES = {
    "solo": {
        "name": "Start",
        "price": 39,
        "max_clients": 20,
        "included_users": 1,
        "team_management": False,
        "worker_app": False,
        "csv_team_import": False,
        "csv_client_import": False,
        "recurring_jobs": False,
        "ai_operator": False,
        "approval_queue": False,
        "automation": False,
        "myob_sync": False,
        "myob_addon_available": False,
        "myob_included": False,
        "payroll_workspace": False,
        "enterprise_user_blocks": False,
    },
    "team": {
        "name": "Crew",
        "price": 89,
        "max_clients": 30,
        "included_users": 5,
        "team_management": True,
        "worker_app": True,
        "csv_team_import": True,
        "csv_client_import": False,
        "recurring_jobs": False,
        "ai_operator": False,
        "approval_queue": False,
        "automation": False,
        "myob_sync": False,
        "myob_addon_available": False,
        "myob_included": False,
        "payroll_workspace": False,
        "enterprise_user_blocks": False,
    },
    "pro": {
        "name": "Operator",
        "price": 149,
        "max_clients": 40,
        "included_users": 15,
        "team_management": True,
        "worker_app": True,
        "csv_team_import": True,
        "csv_client_import": True,
        "recurring_jobs": True,
        "ai_operator": True,
        "approval_queue": True,
        "automation": True,
        "myob_sync": False,
        "myob_addon_available": True,
        "myob_included": False,
        "payroll_workspace": False,
        "enterprise_user_blocks": False,
    },
    "enterprise": {
        "name": "Command",
        "price": 299,
        "max_clients": 50,
        "included_users": 50,
        "team_management": True,
        "worker_app": True,
        "csv_team_import": True,
        "csv_client_import": True,
        "recurring_jobs": True,
        "ai_operator": True,
        "approval_queue": True,
        "automation": True,
        "myob_sync": True,
        "myob_addon_available": True,
        "myob_included": True,
        "payroll_workspace": True,
        "enterprise_user_blocks": True,
        "extra_user_block_size": 50,
        "extra_user_block_price": 99,
    },
}

def normalize_plan(plan):
    value = str(plan or "").strip().lower()
    if not value or value in ("null", "undefined", "none"):
        return None
    if value in PLAN_ORDER:
        return value
    return "solo"

def plan_rank(plan):
    normalized = normalize_plan(plan)
    return PLAN_ORDER.get(normalized, 0)

def has_plan_access(current_plan, required_plan):
    return plan_rank(current_plan) >= plan_rank(required_plan)

def get_plan_features(plan):
    normalized = normalize_plan(plan)
    return PLAN_FEATURES.get(normalized, PLAN_FEATURES["solo"])

def get_plan_display_name(plan):
    normalized = normalize_plan(plan)
    return PLAN_DISPLAY_NAMES.get(normalized, "No plan")

def can_use_feature(plan, feature_key):
    return bool(get_plan_features(plan).get(feature_key, False))

def get_max_clients(plan):
    return int(get_plan_features(plan).get("max_clients", 20))

def get_included_users(plan):
    return int(get_plan_features(plan).get("included_users", 1))

# CHURVOX_TOP_TIER_BACKEND_ENDPOINTS_PLAN_RULES_HOOK_20260528
# Fallback startup hook: server.py imports this module, so retry route registration
# after server.py has finished creating app/db/dependencies.
def _churvox_try_register_top_tier_routes(attempt=0):
    try:
        import importlib
        server = importlib.import_module("server")
        routes = importlib.import_module("churvox_top_tier_routes")
        if routes.register(server):
            return
    except Exception as exc:
        if attempt >= 12:
            print("CHURVOX_TOP_TIER_PLAN_RULES_HOOK_FAILED", repr(exc))
            return
    try:
        import threading
        threading.Timer(0.25, _churvox_try_register_top_tier_routes, kwargs={"attempt": attempt + 1}).start()
    except Exception as exc:
        print("CHURVOX_TOP_TIER_PLAN_RULES_TIMER_FAILED", repr(exc))

_churvox_try_register_top_tier_routes()

File: frontend/src/utils/planRules.js
export const PLAN_ORDER = {
  solo: 1,
  team: 2,
  pro: 3,
  enterprise: 4,
};

export const EXTRA_USER_BLOCK_SIZE = 50;
export const EXTRA_USER_BLOCK_PRICE = 100;

export const normalizePlan = (plan) => {
  const value = String(plan || "").trim().toLowerCase();
  if (value === "solo") return "solo";
  if (value === "team") return "team";
  if (value === "pro") return "pro";
  if (value === "enterprise") return "enterprise";
  if (!value || value === "null" || value === "undefined" || value === "none") return "none";
  return "solo";
};

export const getExtraUserBlocks = (userOrBilling = {}) => {
  const raw = userOrBilling?.extra_user_blocks ?? userOrBilling?.user_blocks ?? userOrBilling?.capacity_blocks ?? userOrBilling?.additional_user_blocks ?? 0;
  const blocks = Number(raw || 0);
  return Number.isFinite(blocks) && blocks > 0 ? Math.floor(blocks) : 0;
};

export const hasPlanAccess = (currentPlan, requiredPlan) => {
  const current = normalizePlan(currentPlan);
  const required = normalizePlan(requiredPlan);
  return (PLAN_ORDER[current] || 0) >= (PLAN_ORDER[required] || 0);
};

export const PLAN_FEATURES = {
  solo: {
    name: "Solo",
    maxClients: 20,
    teamManagement: false,
    csvTeamImport: false,
    csvClientImport: false,
    recurringJobs: false,
    myobSync: false,
    enterpriseUserBlocks: false,
    includedUsers: 1,
  },
  team: {
    name: "Team",
    maxClients: 30,
    teamManagement: true,
    csvTeamImport: true,
    csvClientImport: false,
    recurringJobs: false,
    myobSync: false,
    enterpriseUserBlocks: false,
    includedUsers: 5,
  },
  pro: {
    name: "Pro",
    maxClients: 40,
    teamManagement: true,
    csvTeamImport: true,
    csvClientImport: true,
    recurringJobs: true,
    myobSync: false,
    enterpriseUserBlocks: false,
    includedUsers: 20,
  },
  enterprise: {
    name: "Enterprise",
    maxClients: 50,
    teamManagement: true,
    csvTeamImport: true,
    csvClientImport: true,
    recurringJobs: true,
    myobSync: true,
    enterpriseUserBlocks: true,
    includedUsers: 50,
    extraUserBlockSize: EXTRA_USER_BLOCK_SIZE,
    extraUserBlockPrice: EXTRA_USER_BLOCK_PRICE,
  },
};

export const getPlanFeatures = (plan, userOrBilling = null) => {
  const base = PLAN_FEATURES[normalizePlan(plan)] || PLAN_FEATURES.solo;
  const extraBlocks = getExtraUserBlocks(userOrBilling || {});
  const extraUsers = base.enterpriseUserBlocks ? extraBlocks * EXTRA_USER_BLOCK_SIZE : 0;
  return {
    ...base,
    extraUserBlocks: extraBlocks,
    extraUsers,
    includedUsersBase: base.includedUsers,
    includedUsers: Number(base.includedUsers || 1) + extraUsers,
    extraUserBlockSize: base.extraUserBlockSize || EXTRA_USER_BLOCK_SIZE,
    extraUserBlockPrice: base.extraUserBlockPrice || EXTRA_USER_BLOCK_PRICE,
  };
};

export const canUseFeature = (plan, featureKey) => {
  const features = getPlanFeatures(plan);
  return !!features[featureKey];
};

export const getMaxClients = (plan) => {
  return getPlanFeatures(plan).maxClients;
};
